File: src/algorithmic_toolbox/week5/primitive_calc.py
from typing import Callable, List


IntOper = Callable[[int], int]


def best_opers(target: int, opers: List[IntOper]) -> List[IntOper]:
    table = [None] * (target + 1)
    table[1] = []  # starts at 1 - base case

    for i in range(1, target + 1):
        if table[i] is not None:
            for oper in opers:
                res = oper(i)
                potential_opers = [*table[i], oper]
                if res <= target and (table[res] is None or len(table[res]) > len(potential_opers)):
                    table[res] = potential_opers

    return table[target]


if __name__ == '__main__':
    opers = [
        lambda x: x + 1,
        lambda x: x * 2,
        lambda x: x * 3
    ]

    target = int(input())
    operations = best_opers(target, opers)
    
    print(len(operations))
    
    result = 1
    for oper in operations:
        print(result, end=' ')
        result = oper(result)
    print(result)

File: src/algorithmic_toolbox/week5/PrimitiveCalc.java
package algorithmic_toolbox.week5;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.function.IntUnaryOperator;


public class PrimitiveCalc {
  
  @SuppressWarnings("unchecked")
  public static List<IntUnaryOperator> bestOpers(int target, IntUnaryOperator[] opers) {
    List<IntUnaryOperator>[] table = (List<IntUnaryOperator>[]) (new List[target + 1]);
    table[1] = new ArrayList<IntUnaryOperator>();

    int res;
    List<IntUnaryOperator> potentialOpers;
    for (int i = 0; i <= target; i++) {
      if (table[i] != null) {
        for (IntUnaryOperator oper : opers) {
          res = oper.applyAsInt(i);
          potentialOpers = new ArrayList<>(table[i]);
          potentialOpers.add(oper);

          if (res <= target && (table[res] == null || table[res].size() > potentialOpers.size())) {
            table[res] = potentialOpers;
          }
        }
      }
    }

    return table[target];
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int target = scanner.nextInt();
    scanner.close();

    IntUnaryOperator[] opers = {
      (int x) -> x + 1,
      (int x) -> x * 2,
      (int x) -> x * 3
    };

    List<IntUnaryOperator> opersList = bestOpers(target, opers);

    System.out.println(opersList.size());
    
    int result = 1;
    for (IntUnaryOperator oper : opersList) {
      System.out.printf("%d ", result);
      result = oper.applyAsInt(result);
    }
    System.out.println(result);
  }
}
